source/guides/maintenance/custom-app-container.rst

```diff
@@ -0,0 +1,402 @@
+
+#.	The specific features are built in the first stage via
+	:command:`swupd os-install … bundles=httpd…` with all files from its base layer removed.
+
+#.	The target is a composition of its base layer and the specific feature layer,
+	via :command:`FROM clearlinux/<base layer>:latest`, such as: os-core, httpd,
+	and via :command:`COPY --from=builder /install_root /`.
+
+Using this method, the target container images are kept up to date without
+file duplication. For application-centric containers, `os-core-update` is
+excluded to improve size optimization.
+
+Multi-stage Dockerfiles contain more than one :command:`FROM` directive. All
+of the multi-stage Clear Linux OS Dockerfiles share a common base layer called
+:command:`clearlinux/os-core:latest`. All of the higher level layers inherit
+the Clear Linux OS version from this base layer.
+
+Examples
+********
+
+This section describes the following examples for customizing an application image
+Dockerfile:
+
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Add a bundle
+============
+
+To add new software to a container image, you need to find out which |CL|
+bundle includes it, using the :command:`swupd search <feature>` command.
+(Refer to `swupd source documentation`_ on GitHub for command details.)
+
+After finding out the right bundle(s), update the Dockerfile to include the
+bundle into the published |CL| container. In this example, we chose
+to add :command:`wget` to the clearlinux/redis Dockerfile.
+
+To demonstrate this, a diff of a modified clearlinux/redis Dockerfile is
+provided below to include :command:`wget`.
+
+.. code-block:: bash
+
+	diff --git a/redis/Dockerfile b/redis/Dockerfile
+	index af977cb..b1effab 100644
+	--- a/redis/Dockerfile
+	+++ b/redis/Dockerfile
+	@@ -15,7 +15,7 @@ RUN source /os-release && \
+	     mkdir /install_root \
+	     && swupd os-install -V ${VERSION_ID} \
+	     --path /install_root --statedir /swupd-state \
+	-    --bundles=redis-native,findutils,su-exec --no-boot-update
+	+    --bundles=redis-native,findutils,su-exec,wget --no-boot-update
+
+	 # For some Host OS configuration with redirect_dir on,
+	 # extra data are saved on the upper layer when the same
+
+
+	$ docker build --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy -t clearlinux/redis:wget_added redis/
+
+
+	$ docker run clearlinux/redis:wget_added wget --version
+	GNU Wget 1.20.3 built on linux-gnu.
+
+	-cares +digest -gpgme +https +ipv6 -iri +large-file -metalink +nls
+	-ntlm +opie -psl +ssl/openssl
+	...
+
+Change Clear Linux OS version
+==============================
+
+
+Sometimes during development and debugging, you may need to run experiments
+using |CL| containers built from an older version of the OS. This section
+describes how to accomplish this using the published Clear Linux OS
+Dockerfiles, with different steps for single-stage and multi-stage
+Dockerfiles.
+
+
+Single-stage Clear Linux OS Dockerfiles
+---------------------------------------
+

```

Comment:
Separate action from description. Create discrete steps for user action. Otherwise, move description to a para or a note.